cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1909,6 +1912,145 @@ func maybeMountSave(disk disks.Disk, rootdir string, encrypted bool, mountOpts *
 	return true, nil
 }
 
+func createKernelMounts(runWritableDataDir, kernelName string, rev snap.Revision, isClassic bool) (bool, error) {
+	driversStandardDir := kernel.DriversTreeDir(runWritableDataDir, kernelName, rev)
+	// On UC first boot the drivers dir is initially under
+	// _writable_defaults, so we need to check that directory too. But the
+	// mount happens after handle-writable-paths has run, so the units
+	// mounting /lib/{modules,firmware} can use driversStandardDir always.
+	driversFirstBootDir := kernel.DriversTreeDir(
+		filepath.Join(runWritableDataDir, "_writable_defaults"), kernelName, rev)
+	var driversDir string
+	switch {
+	case osutil.IsDirectory(driversStandardDir):
+		driversDir = driversStandardDir
+	case osutil.IsDirectory(driversFirstBootDir):
+		driversDir = driversFirstBootDir
+	default:
+		logger.Noticef("no drivers tree at %s", driversStandardDir)
+		return false, nil
+	}
+	logger.Noticef("drivers tree found in %s", driversDir)
+
+	// 1. Mount unit for the kernel snap
+	cpi := snap.MinimalSnapContainerPlaceInfo(kernelName, rev)
+	squashfsPath := filepath.Join(runWritableDataDir, dirs.StripRootDir(cpi.MountFile()))
+	// snapRoot is where we will find the /snap directory where
+	// snaps/components will be mounted
+	snapRoot := filepath.Join("writable", "system-data")
+	if isClassic {
+		snapRoot = "sysroot"
+	}
+	where := filepath.Join(dirs.GlobalRootDir, snapRoot, dirs.StripRootDir(cpi.MountDir()))
+	if err := writeInitramfsMountUnit(squashfsPath, where, squashfsUnit); err != nil {
+		return false, err
+	}
+
+	// 2. Mount units for kernel-modules components
+	if err := createKernelModulesMountUnits(
+		runWritableDataDir, snapRoot, driversDir, kernelName); err != nil {
+		return false, err
+	}
+
+	// 3. Mount units for /lib/{modules,firmware}
+	for _, subDir := range []string{"modules", "firmware"} {
+		what := filepath.Join(driversStandardDir, "lib", subDir)
+		where := filepath.Join(dirs.GlobalRootDir, "sysroot", "usr", "lib", subDir)
+		if err := writeInitramfsMountUnit(what, where, bindUnit); err != nil {
+			return false, fmt.Errorf("while creating mount for %s in %s: %v",
+				what, where, err)
+		}
+	}
+
+	// daemon-reload is not needed because it is done from initramfs
+	// later, this happens because
+	// 1. On UC /etc/fstab is changed and systemd's
+	//    initrd-parse-etc.service does the reload
```

Comment:
In recent version of systemd, the `daemon-reload` will happen only if `/etc/fstab` contains things related to initrd. Mainly `x-initrd.mount` option. So maybe it would be nice to say `/etc/fstab` contains those.

Reply:
I've clarified in the comment, thanks